File: python/interview/classic/semaphore/semaphore.py
"""手写计数信号量：mutex + Condition（与 threading.Semaphore 对照学习）。"""

from __future__ import annotations

import threading
import time
from typing import Optional


class Semaphore:
    def __init__(self, value: int) -> None:
        if value < 0:
            raise ValueError("value must be non-negative")
        self._cond = threading.Condition()
        self._count = value

    def acquire(self, blocking: bool = True, timeout: Optional[float] = None) -> bool:
        with self._cond:
            if not blocking:
                if self._count == 0:
                    return False
                self._count -= 1
                return True
            ok = self._cond.wait_for(lambda: self._count > 0, timeout=timeout)
            if not ok:
                return False
            self._count -= 1
            return True

    def release(self, n: int = 1) -> None:
        if n < 1:
            raise ValueError("n must be >= 1")
        with self._cond:
            self._count += n
            for _ in range(n):
                self._cond.notify()


if __name__ == "__main__":
    sem = Semaphore(2)
    gate = threading.Barrier(3)
    results: list[int] = []
    lock = threading.Lock()

    def worker(i: int) -> None:
        gate.wait()
        assert sem.acquire(timeout=2.0)
        with lock:
            results.append(i)
        time.sleep(0.05)
        sem.release()

    ts = [threading.Thread(target=worker, args=(i,)) for i in range(3)]
    for t in ts:
        t.start()
    for t in ts:
        t.join()
    assert len(results) == 3
    sem2 = Semaphore(0)
    assert not sem2.acquire(blocking=False)
    sem2.release()
    assert sem2.acquire(blocking=False)
    print("semaphore OK")

File: cpp/interview/classic/semaphore/semaphore.cpp
// 手写计数信号量：mutex + condition_variable（C++17）
#include <bits/stdc++.h>
#include <cassert>
#include <chrono>
#include <condition_variable>
#include <mutex>
#include <thread>
using namespace std;

class Semaphore {
public:
    explicit Semaphore(int value) : count_(value) {
        if (value < 0) throw invalid_argument("value>=0");
    }

    void acquire() {
        unique_lock<mutex> lk(mtx_);
        cv_.wait(lk, [&] { return count_ > 0; });
        --count_;
    }

    bool try_acquire() {
        lock_guard<mutex> lk(mtx_);
        if (count_ == 0) return false;
        --count_;
        return true;
    }

    void release(int n = 1) {
        if (n < 1) throw invalid_argument("n>=1");
        lock_guard<mutex> lk(mtx_);
        count_ += n;
        for (int i = 0; i < n; ++i) cv_.notify_one();
    }

private:
    mutex mtx_;
    condition_variable cv_;
    int count_;
};

int main() {
    Semaphore sem(2);
    mutex pk;
    atomic<int> cur{0};
    int peak = 0;

    auto worker = [&] {
        sem.acquire();
        int c = ++cur;
        {
            lock_guard<mutex> lk(pk);
            peak = max(peak, c);
        }
        this_thread::sleep_for(chrono::milliseconds(5));
        --cur;
        sem.release();
    };

    vector<thread> ts;
    for (int i = 0; i < 3; ++i) ts.emplace_back(worker);
    for (auto& t : ts) t.join();
    assert(peak <= 2);

    Semaphore z(0);
    assert(!z.try_acquire());
    z.release();
    assert(z.try_acquire());
    cout << "semaphore OK" << endl;
    return 0;
}
